File: stats.js
// use dotenv file
require('dotenv').config();

// create a function that queries the github graphql api for the user's stats
const { GraphQLClient } = require('graphql-request');
const GITHUB_API_URL = "https://api.github.com/graphql"

const client = new GraphQLClient(GITHUB_API_URL, {
  headers: {
    Authorization: `bearer ${process.env.GITHUB_TOKEN}`,
  },
});


const getStats = async (username) => {
  const query = `
    {
        user(login: "${username}") {
          contributionsCollection(from: "2023-01-01T00:00:00") {
            totalCommitContributions
            issueContributions {
              totalCount
            }
          }
          followers {
            totalCount
          }
          following {
            totalCount
          }
        }
      }
    `;

  const data = await client.request(query);
  return data;
}

const getStargazers = async (username, afterCursor) => {
  var repos = [];
  const query = `
    {
      user(login: "${username}") {
          repositories(
          ownerAffiliations: [COLLABORATOR,OWNER]
          first: 5
          privacy: PUBLIC
          ${afterCursor ? `after: "${afterCursor}"` : ""}
          ) {
          nodes {
              stargazerCount
              owner {
              login
              }
              name
          }
          pageInfo {
              endCursor
          }
          }
      }
      }
    `;


  const data = await client.request(query);
  const endCursor = data.user.repositories.pageInfo.endCursor;
  repos = data.user.repositories.nodes.filter(repo => repo.stargazerCount > 0)

  if (endCursor) {
    repos = [...repos, ...await getStargazers(username, endCursor)]
  }
  return repos;
}


const collectStats = async (username) => {
  const stargazerDetails = await getStargazers(username);
  const statsUser = await getStats(username);

  const stats = {
    commitContributions: statsUser.user.contributionsCollection.totalCommitContributions,
    issueContributions: statsUser.user.contributionsCollection.issueContributions.totalCount,
    followers: statsUser.user.followers.totalCount,
    stargazerDetails: stargazerDetails,
    stargazerSum: stargazerDetails.reduce((acc, repo) => acc + repo.stargazerCount, 0)
  }

  return stats


}

export default collectStats;

File: stats.py
from pprint import pprint
import requests
from jinja2 import Environment

    

query_issues_followers = """
{
  user(login: "bart6114") {
    contributionsCollection(from: "2023-01-01T00:00:00") {
      totalCommitContributions
      issueContributions {
        totalCount
      }
    }
    followers {
      totalCount
    }
    following {
      totalCount
    }
  }
}
"""

r = requests.post("https://api.github.com/graphql", 
json = {'query': query_issues_followers}, 
headers={"Authorization": "bearer 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"})
pprint(r.json())
exit()



def fetch_repos(repos_data: list = [], after_cursor: str = None):
    env = Environment()
    query = env.from_string("""
    {
    user(login: "murilo-cunha") {
        repositories(
        ownerAffiliations: [COLLABORATOR,OWNER]
        first: 5
        privacy: PUBLIC
        {% if after_cursor %}
        after: "{{ after_cursor }}"
        {% endif %}
        ) {
        nodes {
            stargazerCount
            owner {
            login
            }
            name
        }
        pageInfo {
            endCursor
        }
        }
    }
    }
    """).render({"after_cursor": after_cursor})

    r = requests.post("https://api.github.com/graphql", 
    json = {'query': query}, 
    headers={"Authorization": "bearer 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"})
    res = r.json()

    repos_data = res["data"]["user"]["repositories"]["nodes"]
    end_cursor  = res["data"]["user"]["repositories"]["pageInfo"]["endCursor"]

    if end_cursor is not None: 
        repos_data = repos_data + fetch_repos(repos_data, end_cursor)
    
    return repos_data
  

res = fetch_repos()

t = {}
for r in res:
  if r["stargazerCount"] > 0:
    if not r["owner"]["login"] in t:
      t[r["owner"]["login"]] = [r]
    else:
      t[r["owner"]["login"]].append(r)

pprint(t)

muris_stars = 4
total_stars = sum([x['stargazerCount'] for x in res])

print(f"In total this gives you {total_stars} stars")




